flowbench/BenchZstd.cpp

```diff
@@ -0,0 +1,242 @@
+
+// Benchmark zstd performance. To use it:
+//  # export BM_ZSTD_DATA=path/to/datafile
+//  # bin/flowbench --benchmark_filter=bench_zstd
+
+// Compress with raw ZSTD API
+static inline std::string compress(const std::string& data, int level) {
+	const char* srcBegin = data.data();
+	size_t destSize = ZSTD_compressBound(data.size());
+	std::unique_ptr<char[]> dest = std::make_unique<char[]>(destSize);
+	size_t bytes = ZSTD_compress(dest.get(), destSize, srcBegin, data.size(), level);
+	return std::string(dest.get(), bytes);
+}
+
+// Compress with raw ZSTD API ZSTD_compress2
+static inline std::string compress2(const std::string& data, ZSTD_CCtx* cctx) {
+	const char* srcBegin = data.data();
+	size_t destSize = ZSTD_compressBound(data.size());
+	std::unique_ptr<char[]> dest = std::make_unique<char[]>(destSize);
+	size_t bytes = ZSTD_compress2(cctx, dest.get(), destSize, srcBegin, data.size());
+	return std::string(dest.get(), bytes);
+}
+
+// Deompress with raw ZSTD API
+static inline std::string decompress(const std::string& data) {
+	size_t destSize = ZSTD_decompressBound(data.data(), data.size());
+	std::unique_ptr<char[]> dest = std::make_unique<char[]>(destSize);
+	size_t bytes = ZSTD_decompress(dest.get(), destSize, data.data(), data.size());
+	return std::string(dest.get(), bytes);
+}
+
+// Compress with raw ZSTD stream API
+static inline std::string compressAsStream(const std::string& data, ZSTD_CStream* cstream) {
+	size_t dstSize = ZSTD_compressBound(data.size());
+	std::unique_ptr<char[]> dst = std::make_unique<char[]>(dstSize);
+
+	ZSTD_inBuffer in;
+	in.src = data.data();
+	in.size = data.size();
+	in.pos = 0;
+	ZSTD_outBuffer out;
+	out.dst = dst.get();
+	out.size = dstSize;
+	out.pos = 0;
+
+	ZSTD_compressStream(cstream, &out, &in);
+	ZSTD_flushStream(cstream, &out);
+	return std::string(dst.get(), out.pos);
+}
+
+// Decompress with raw ZSTD stream API
+static inline std::string decompressAsStream(const std::string& data, ZSTD_DStream* dstream) {
+	size_t destSize = ZSTD_decompressBound(data.data(), data.size());
+	std::unique_ptr<char[]> dest = std::make_unique<char[]>(destSize);
+
+	ZSTD_inBuffer in;
+	in.src = data.data();
+	in.size = data.size();
+	in.pos = 0;
+	ZSTD_outBuffer out;
+	out.dst = dest.get();
+	out.size = destSize;
+	out.pos = 0;
+
+	do {
+		ZSTD_decompressStream(dstream, &out, &in);
+	} while (in.pos < in.size && out.pos < out.size);
+
+	return std::string(dest.get(), destSize);
+}
+
+// Generate uncompressed data for compression testing
+static std::string genUncompressedData() {
+	char* dataFileName = std::getenv("BM_ZSTD_DATA");
+	if (dataFileName) {
+		std::ifstream file(dataFileName, std::ios::binary | std::ios::ate);
+		std::streamsize size = file.tellg();
+		file.seekg(0, std::ios::beg);
+		std::string buf(size, ' ');
+		file.read(buf.data(), size);
+		std::printf("Load test data %s: %ld bytes\n", dataFileName, size);
+		return buf;
+	} else {
+		return deterministicRandom()->randomAlphaNumeric(1048576);
```

Comment:
For consistency across runs, it might make sense to generate a `DeterministicRandom` object from a seed fixed at compile-time, so that every run uses the same "random" data. Then again, maybe it is worthwhile getting the benchmark to run against a variety of seeds.

Reply:
good idea. fixed. thank you!